File: tests/bench/dispatch.go
// The same work as `dispatch.skuld`, in Go. See `README.md`.
package main

import (
	"fmt"
	"os"
	"strconv"
)

type Shape interface {
	Area() int64
}

type Square struct{ side int64 }
type Rectangle struct{ width, height int64 }
type Triangle struct{ base, height int64 }

func (s *Square) Area() int64    { return s.side * s.side }
func (r *Rectangle) Area() int64 { return r.width * r.height }
func (t *Triangle) Area() int64  { return t.base * t.height / 2 }

func main() {
	count := 0
	if len(os.Args) > 1 {
		count, _ = strconv.Atoi(os.Args[1])
	}
	shapes := []Shape{}
	for index := 0; index < count; index++ {
		i := int64(index)
		switch index % 3 {
		case 0:
			shapes = append(shapes, &Square{side: i % 100})
		case 1:
			shapes = append(shapes, &Rectangle{width: i % 50, height: i % 20})
		default:
			shapes = append(shapes, &Triangle{base: i % 40, height: i % 30})
		}
	}
	var total int64 = 0
	for _, shape := range shapes {
		total = (total + shape.Area()) % 1000000007
	}
	fmt.Println(total)
}

File: tests/bench/dispatch.rs
// The same work as `dispatch.skuld`, in Rust. See `README.md`.

trait Shape {
    fn area(&self) -> i64;
}

struct Square {
    side: i64,
}
struct Rectangle {
    width: i64,
    height: i64,
}
struct Triangle {
    base: i64,
    height: i64,
}

impl Shape for Square {
    fn area(&self) -> i64 {
        self.side * self.side
    }
}
impl Shape for Rectangle {
    fn area(&self) -> i64 {
        self.width * self.height
    }
}
impl Shape for Triangle {
    fn area(&self) -> i64 {
        self.base * self.height / 2
    }
}

fn main() {
    let count: i64 = std::env::args()
        .nth(1)
        .and_then(|text| text.parse().ok())
        .unwrap_or(0);
    let mut shapes: Vec<Box<dyn Shape>> = Vec::new();
    for index in 0..count {
        match index % 3 {
            0 => shapes.push(Box::new(Square { side: index % 100 })),
            1 => shapes.push(Box::new(Rectangle {
                width: index % 50,
                height: index % 20,
            })),
            _ => shapes.push(Box::new(Triangle {
                base: index % 40,
                height: index % 30,
            })),
        }
    }
    let mut total: i64 = 0;
    for shape in &shapes {
        total = (total + shape.area()) % 1000000007;
    }
    println!("{total}");
}
